File: 2015/09/R/09.R
suppressPackageStartupMessages({
  library(combinat)
})

input <- readLines("stdin")

# generate table of town-to-town distances
input <- as.data.frame(t(mapply(
  input,
  regexec("(\\w+) to (\\w+) = (\\d+)", input),
  FUN = function(i, m) {
    substring(i, m, m + attr(m, "match.length") - 1L)
  }
))[,2:4], row.names = FALSE)
names(input) <- c("x", "y", "d")
input$d <- as.numeric(input$d)
input <- rbind(input, data.frame(x = input$y, y = input$x, d = input$d))

# get unique towns
towns <- unique(input$x)

# generate table of trips
trips <- do.call(rbind, lapply(permn(towns), function(i) {
    data.frame(x = i[-length(i)], y = i[-1], leg = 1:(length(i)-1))
}))
trips$trip <- rep(1:factorial(length(towns)), each = length(towns) - 1L)
trips <- merge(trips, input, by = c("x", "y"))
trips <- trips[order(trips$trip, trips$leg),]

# aggregate trips by sum of leg distances
trip_dists <- aggregate(d ~ trip, data = trips, sum)
cat(min(trip_dists$d), "\n")
cat(max(trip_dists$d), "\n")

File: 2015/09/julia/09.jl
using Combinatorics

towns = Set{String}()
leg_dists = Dict{Tuple,Int}()

map(readlines(stdin)) do line
    m = match(r"(\w+) to (\w+) = (\d+)", line)
    union!(towns, [m[1], m[2]])
    leg_dists[(m[2], m[1])] = leg_dists[(m[1], m[2])] = parse(Int, m[3])
end

trip_dists = map(permutations(collect(towns))) do stops
    sum(leg_dists[leg] for leg in zip(stops[1:end-1], stops[2:end]))
end

println(minimum(trip_dists))
println(maximum(trip_dists))
